trieste/objectives/multi_objectives.py

```diff
@@ -236,3 +237,76 @@ def gen_pareto_optimal_points(n: int, seed: int | None = None) -> TensorType:
         search_space=Box([0.0], [1.0]) ** d,
         gen_pareto_optimal_points=gen_pareto_optimal_points,
     )
+
+
+@dataclass(frozen=True)
+class ConstrainedMultiObjectiveTestProblem(MultiObjectiveTestProblem):
+    """
+    Convenience container class for synthetic constrained multi-objective test functions, containing
+    a generator for the pareto optimal points, which can be used as a reference of performance
+    measure of certain multi-objective optimization algorithms.
+    """
+
+    constraint: Callable[[TensorType], TensorType]
+    """The synthetic test function's constraints"""
+
+
+def ConstrainedBraninCurrin() -> ConstrainedMultiObjectiveTestProblem:
+    """
+    The ConstrainedBraninCurrin problem, typically evaluated over :math:`[0, 1]^2`.
+    See :cite:`belakaria2019max` and :cite:`daulton2020differentiable`
+    (the latter for adding the constraint) for details.
+
+    :return: The problem specification.
+    """
+
+    def gen_pareto_optimal_points(n: int, seed: int | None = None) -> TensorType:
+        return tf.zeros(shape=0)
+
+    def evaluate_slack_true(x: TensorType) -> TensorType:
+        """
+        The constraint of branincurrin problem, < 0 is feasible
+
+        :param x: The points at which to evaluate the function, with shape [..., d].
+        :raise ValueError (or InvalidArgumentError): If ``x`` has an invalid shape.
+        """
+        x = x * (
+            tf.constant([10.0, 15.0], dtype=x.dtype) - tf.constant([-5.0, 0.0], dtype=x.dtype)
+        ) + tf.constant([-5.0, 0.0], dtype=x.dtype)
+        return (x[..., :1] - 2.5) ** 2 + (x[..., 1:] - 7.5) ** 2 - 50
+
+    return ConstrainedMultiObjectiveTestProblem(
+        name="ConstrainedBraninCurrin",
+        objective=branin_currin,
+        constraint=evaluate_slack_true,
+        search_space=Box([0.0], [1.0]) ** 2,
+        gen_pareto_optimal_points=gen_pareto_optimal_points,
+    )
+
+
+def branin_currin(x: TensorType) -> TensorType:
+    """
+    The branincurrin synthetic function.
+
+    :param x: The points at which to evaluate the function, with shape [..., d].
+    :raise ValueError (or InvalidArgumentError): If ``x`` has an invalid shape.
+    """
+    tf.debugging.assert_shapes([(x, (..., 2))])
+    return tf.concat([branin(x), currin(x)], axis=-1)
```

Comment:
mixing objectives and constraints seems like a wrong design, I think it makes sense to keep them separate, since constraints typically have different treatment (e.g. computing prob of feasibility)

what I would suggest to return `currin` as constraint and add another method to the interface for evaluating the constraint (i.e. `evaluate_slack_true`), we could call it `evaluate_constraint`? 

@uri-granta what do you think?

Reply:
Sorry that i have failed to make this clear: for the `branincurrin` problem the objective function is a concatenation of `[branin, currin]`, so both are objectives. For the constraint we indeed want to have a separate function, i.e.,  `evaluate_constraint` as suggested.

Reply:
ah, ok, clearer now, thanks!